File: csharp/AppEncryption/AppEncryption/Envelope/KeyMeta.cs
using System;
using GoDaddy.Asherah.AppEncryption.Util;
using Newtonsoft.Json.Linq;

namespace GoDaddy.Asherah.AppEncryption.Envelope
{
    /// <summary>
    /// The <see cref="KeyMeta"/> format is defined as:
    /// <code>
    /// {
    ///   KeyId: "some_key_id",
    ///   Created: 1534553054
    /// }
    /// </code>
    /// </summary>
    public class KeyMeta
    {
        /// <summary>
        /// Initializes a new instance of the <see cref="KeyMeta"/> class using the provided parameters. Key meta is the
        /// metadata in <see cref="EnvelopeKeyRecord"/> that references a parent key in the key hierarchy.
        /// Note that for system keys, this content may be embedded within the encrypted key content, depending on the
        /// KMS being used.
        /// </summary>
        ///
        /// <param name="keyId">The key id.</param>
        /// <param name="created">The creation time of the key.</param>
        public KeyMeta(string keyId, DateTimeOffset created)
        {
            KeyId = keyId;
            Created = created;
        }

        /// <summary>
        /// Initializes a new instance of the <see cref="KeyMeta"/> class from the source json.
        /// </summary>
        ///
        /// <param name="sourceJson">Source json to create the key meta.</param>
        public KeyMeta(Json sourceJson)
        {
            KeyId = sourceJson.GetString("KeyId");
            Created = sourceJson.GetDateTimeOffset("Created");
        }

        internal string KeyId { get; }

        internal virtual DateTimeOffset Created { get; }

        /// <summary>
        /// Converts the <see cref="KeyMeta"/> to a <see cref="JObject"/> with below format:
        /// <code>
        /// {
        ///   "KeyId": "some_key_id",
        ///   "Created": 1534553054
        /// }
        /// </code>
        /// </summary>
        ///
        /// <returns>The <see cref="KeyMeta"/> converted to a <see cref="JObject"/> object.</returns>
        public JObject ToJson()
        {
            Json json = new Json();
            json.Put("KeyId", KeyId);
            json.Put("Created", Created);
            return json.ToJObject();
        }

        /// <inheritdoc/>
        public override bool Equals(object obj)
        {
            if (this == obj)
            {
                return true;
            }

            if (obj == null || GetType() != obj.GetType())
            {
                return false;
            }

            KeyMeta other = (KeyMeta)obj;
            return Equals(KeyId, other.KeyId)
                   && Equals(Created, other.Created);
        }

        /// <inheritdoc/>
        public override int GetHashCode()
        {
            return (KeyId, Created).GetHashCode();
        }

        /// <inheritdoc/>
        public override string ToString()
        {
            return "KeyMeta [KeyId=" + KeyId + ", Created=" + Created + "]";
        }
    }
}

File: java/app-encryption/src/main/java/com/godaddy/asherah/appencryption/envelope/KeyMeta.java
package com.godaddy.asherah.appencryption.envelope;

import java.time.Instant;
import java.util.Objects;

import org.json.JSONObject;

import com.godaddy.asherah.appencryption.utils.Json;

/**
 * The {@code KeyMeta} format is below.
 * <pre>
 * {
 *   KeyId: "some_key_id",
 *   Created: 1534553054
 * }
 * </pre>
 */
public class KeyMeta {
  private final String keyId;
  private final Instant created;

  /**
   * Creates a new {@code KeyMeta} instance using the provided parameters. {@code KeyMeta} is the metadata in
   * {@link EnvelopeKeyRecord} that references a parent key in the key hierarchy.
   * Note that for system keys, this content may be embedded within the encrypted key content, depending on the KMS
   * being used.
   *
   * @param keyId The key Id.
   * @param created The creation time of the key.
   */
  public KeyMeta(final String keyId, final Instant created) {
    this.keyId = keyId;
    this.created = created;
  }

  KeyMeta(final Json sourceJson) {
    keyId = sourceJson.getString("KeyId");
    created = sourceJson.getInstant("Created");
  }

  /**
   * Converts the {@code KeyMeta} to a {@link org.json.JSONObject} with below format.
   * <pre>
   * {
   *   "KeyId": "some_key_id",
   *   "Created": 1534553054
   * }
   * </pre>
   *
   * @return The {@code KeyMeta} converted to a {@link org.json.JSONObject} object.
   */
  public JSONObject toJson() {
    Json json = new Json();
    json.put("KeyId", keyId);
    json.put("Created", created);
    return json.toJsonObject();
  }

  public String getKeyId() {
    return keyId;
  }

  public Instant getCreated() {
    return created;
  }

  @Override
  public int hashCode() {
    return Objects.hash(keyId, created);
  }

  @Override
  public boolean equals(final Object obj) {
    if (this == obj) {
      return true;
    }
    if (obj == null || getClass() != obj.getClass()) {
      return false;
    }

    KeyMeta other = (KeyMeta) obj;
    return Objects.equals(keyId, other.keyId)
        && Objects.equals(created, other.created);
  }

  @Override
  public String toString() {
    return "KeyMeta [keyId=" + keyId + ", created=" + created + "]";
  }
}
